src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,696 @@
+#include "seastar/net/inet_address.hh"
+#include "seastar/core/reactor.hh"
+#include "seastar/core/thread.hh"
+#include "seastar/util/std-compat.hh"
+#include <boost/algorithm/string.hpp>
+
+/**
+  A Crimson-wise version of the admin socket - implementation file
+
+  \todo handle the unlinking of the admin socket. Note that 'cleanup_files'
+  at-exit functionality is not yet implemented in Crimson.
+*/
+
+namespace {
+seastar::logger& logger()
+{
+  return crimson::get_logger(ceph_subsys_osd);
+}
+}  // namespace
+
+namespace crimson::admin {
+
+/**
+  Hooks table - iterator support
+ */
+
+AdminHooksIter AdminSocket::begin()
+{
+  AdminHooksIter it{ *this };
+  return it;
+}
+
+AdminHooksIter AdminSocket::end()
+{
+  AdminHooksIter it{ *this, true };
+  return it;
+}
+
+/**
+  note that 'end-flag' is an optional parameter, and is only used internally (to
+  signal the end() state).
+ */
+AdminHooksIter::AdminHooksIter(AdminSocket& master, bool end_flag)
+    : m_master{ master }, m_end_marker{ end_flag }
+{
+  if (end_flag) {
+    // create the equivalent of servers.end();
+    m_miter = m_master.servers.end();
+  } else {
+    m_miter = m_master.servers.begin();
+    m_siter = m_miter->second.m_hooks.begin();
+  }
+}
+
+AdminHooksIter& AdminHooksIter::operator++()
+{
+  if (!m_end_marker) {
+    ++m_siter;
+    if (m_siter == m_miter->second.m_hooks.end()) {
+      // move to the next server-block
+      m_miter++;
+      if (m_miter == m_master.servers.end()) {
+        m_end_marker = true;
+        return *this;
+      }
+
+      m_siter = m_miter->second.m_hooks.begin();
+    }
+  }
+  return *this;
+}
+
+/**
+  the default implementation of the hook API
+
+  Note that we never throw or return a failed future.
+ */
+seastar::future<bool> AdminSocketHook::call(std::string_view  command,
+                                            const cmdmap_t&   cmdmap,
+                                            std::string_view  format,
+                                            ceph::bufferlist& out) const
+{
+  unique_ptr<Formatter> f{ Formatter::create(format, "json-pretty"sv,
+                                             "json-pretty"s) };
+
+  // customize the output section, as required by a couple of hook APIs:
+  std::string section{ section_name(command) };
+  boost::replace_all(section, " ", "_");
+  if (format_as_array()) {
+    f->open_array_section(section.c_str());
+  } else {
+    f->open_object_section(section.c_str());
+  }
+
+  bool bres{ false };  ///< 'false' for any failure of in API execution
+
+  /*
+    call the command-specific hook.
+    A note re error handling:
+        - will be modified to use the new 'erroretor'. For now:
+        - exec_command() may throw or return an exceptional future. We return an
+           error message on both failure scenarios.
+   */
+  return seastar::do_with(
+    std::move(f), std::move(bres),
+    [this, &command, &cmdmap, &format, &out](unique_ptr<Formatter>& ftr,
```

Comment:
i'd suggest avoid encoding the type into the variable name. the name should explain what it represents, not how it represents.

Reply:
Agreed. Fixed for all ml comments that were not meant for Doxygen.
(This specific one is a file comment for DX)

Reply:
? 
I used 'ftr' as a shorthand for formatter - the functionality, not the type.

Reply:
i am referencing "bres" also. and could you use `formatter` or `fmt` instead?